### PR TITLE
Add UI tests for Sources/Integrations with rich Playwright reporting (FLPATH-2976)

### PR DESCRIPTION
**Pull request by testetson22**
### Summary

This PR adds comprehensive UI tests for the Sources/Integrations management feature (FLPATH-2976), along with rich Playwright artifact capture and HTML reporting for improved debugging.

**Note: WIP, but not a Draft (so CI runs) until artifact collection is verified in CI**

### Changes

- **New UI tests for Sources/Integrations** (`test_sources.py`)
  - 6 end-to-end workflow tests covering complete user journeys
  - Navigation, empty state, wizard flow, table display, CRUD operations
  - Status validation for integrations

- **Playwright artifact capture** (`conftest.py`)
  - Screenshots: captured for all tests by default
  - Videos: captured only for failures (configurable via `PLAYWRIGHT_VIDEO`)
  - Traces: captured for all tests, viewable at trace.playwright.dev
  - Configurable via environment variables

- **pytest-html integration**
  - Rich HTML report with embedded screenshots and videos
  - Trace download links with file size info
  - Automatic artifact copying to `ARTIFACT_DIR` for CI

- **UI test best practices**
  - Replaced hard `wait_for_timeout()` calls with explicit waits
  - Scoped test cleanup to only delete test-prefixed sources
  - Added specific form locators and assertion messages
  - Documented parallel execution limitations

- **New prompt for CI artifact browsing** (`.cursor/prompts/browse-prow-artifacts.md`)

- **API Test Code Consolidation for UI test setup**
    - `get_fresh_auth_header(keycloak_config, http_session)`**
        - Lightweight alternative for getting just the Authorization header
        - Used when tests need to refresh tokens mid-execution
        - Returns `Optional[Dict[str, str]]` with the header or `None`
    - `create_authenticated_session(keycloak_config, content_type=None)`**
        - Factory function for pre-configured `requests.Session` objects
        - Handles token acquisition, SSL verification, and optional Content-Type
        - Used for module-scoped fixtures and one-off API operation

### Test Plan

- [x] All 6 UI sources tests pass locally
- [x] Artifact capture working (screenshots, traces, videos on failure)
- [x] HTML report generates with embedded artifacts
- [x] Verify artifact collection in CI (`ARTIFACT_DIR` integration)
- [x] Verify report accessibility from Prow artifacts

### Jira

- [FLPATH-2976](https://redhat.atlassian.net/browse/FLPATH-2976) - Sources/Integrations UI management

### CODE REVIEW

File: tests/suites/ui/conftest.py [hunk 1]
```diff
@@ -131,56 +236,125 @@ def authenticated_context(
...
+                print(f"\n⚠️ Failed to rename video: {e}")
+        else:
+            # Delete video for passing tests in retain-on-failure mode
+            try:
+                os.remove(video_path)
+            except Exception:
+                pass
+    
+    # Close context (must happen after trace/video handling)
+    authenticated_context.close()
```
Comment (jordigilh) [inline tests/suites/ui/conftest.py]: **[CRITICAL — Architecture]** Fixture lifecycle ownership violation.

`authenticated_page` closes `authenticated_context` here, but `authenticated_context` is a separate fixture that should own its own lifecycle. This creates a hidden contract between the two fixtures:

1. If a test uses `authenticated_context` directly (without `authenticated_page`), the context is **never closed** — it leaks.
2. The `authenticated_context` fixture's docstring at line 253 explicitly says "Do NOT close context here", which means the ownership responsibility is undocumented except via a comment.

Suggested fix: move trace/video/screenshot teardown logic into `authenticated_context` itself (using `request` which you already accept), and let `authenticated_page` only manage the `page` object. Each fixture should own its own cleanup.

File: tests/suites/ui/test_sources.py [hunk 2]
```diff
@@ -0,0 +1,846 @@
...
+    - Empty state tests require truly empty state to pass
+    - In CI, sources are ephemeral and can be recreated
+    """
+    token = obtain_jwt_token(keycloak_config)
+    session = requests.Session()
+    session.headers["Authorization"] = f"Bearer {token.access_token}"
+    session.headers["Content-Type"] = "application/json"
+    session.verify = False
+
+    response = session.get(f"{gateway_url}/cost-management/v1/sources")
```
Comment (jordigilh) [inline tests/suites/ui/test_sources.py]: **[HIGH — QE]** `ensure_no_sources` deletes **all** sources, not just test-created ones.

This fixture iterates over every source in the namespace and deletes them unconditionally. In shared CI environments, or if IQE tests run in the same namespace and create their own sources, this will destroy sources created by other test suites — causing cascading failures.

The docstring acknowledges this is intentional, but it's risky. Consider:
1. Scope deletion to test-prefixed sources only (matching `TEST_SOURCE_PREFIXES` at line 448)
2. Or ensure these tests run in an isolated namespace dedicated to UI tests
3. At minimum, log which sources are being deleted so failures in other suites can be traced back here

**Review by jordigilh — COMMENTED** [review]
## Readiness Audit — PR #161 (FLPATH-2976)

Multi-persona review covering QE, test quality, security, architecture, SRE, and FedRAMP.

**Summary:** 1 critical, 3 high, 4 medium findings. The critical issue (fixture lifecycle ownership) should be addressed before merge. See inline comments for details.

| Category | Verdict |
|---|---|
| QE / Test Quality | Needs work |
| UX / Accessibility | Acceptable |
| Production Security | Acceptable |
| Architecture | Needs work |
| SRE / Operations | Acceptable |
| FedRAMP | No impact |

File: tests/suites/ui/conftest.py [hunk 3]
```diff
@@ -99,21 +157,68 @@ def keycloak_login_url(keycloak_config: KeycloakConfig) -> str:
...
 def authenticated_context(
     browser: Browser,
     ui_url: str,
     keycloak_config: KeycloakConfig,
+    request,
 ) -> Generator[BrowserContext, None, None]:
     """Create a browser context with authenticated session.
     
     Performs Keycloak login and stores the session for the test.
     Uses test/test credentials by default (configurable via env vars).
```
Comment (jordigilh) [inline tests/suites/ui/conftest.py]: **[HIGH — Security]** Hardcoded default credentials (`test`/`test`).

Lines 232-233 default to `username="test"`, `password="test"` via `os.environ.get(...)`. These are test-environment-only credentials, but since this is committed to the shared repo, please add a comment or docstring explicitly stating:
1. These credentials are only valid in ephemeral CI test environments
2. They must never match any staging/production credentials
3. The Keycloak user is provisioned by the test harness bootstrap

This is a FedRAMP hygiene item — auditors will flag hardcoded credential defaults without explicit documentation of their scope.

File: tests/suites/ui/conftest.py [hunk 4]
```diff
@@ -189,3 +363,228 @@ def pytest_runtest_makereport(item, call):
...
+                f'<img src="data:image/png;base64,{screenshot_data}" '
+                f'alt="Screenshot" style="max-width:100%; border:1px solid #ccc;"/>'
+                f'</div></details>'
+            ))
+        
+        # Add video if exists (only for failed tests - passing tests shouldn't have videos)
+        video_path = os.path.join(_get_videos_dir(), f"{test_name}.webm")
+        if os.path.exists(video_path):
+            with open(video_path, "rb") as f:
+                video_data = base64.b64encode(f.read()).decode()
```
Comment (jordigilh) [inline tests/suites/ui/conftest.py]: **[HIGH — Architecture/SRE]** Base64-encoding video files into the HTML report will produce enormous files.

A 30-second WebM video at 720p can be 5-15MB. Base64 encoding inflates that by ~33%. If multiple tests fail, the HTML report could easily exceed 60MB — making it unusable in browsers and expensive to store in CI artifact buckets.

Suggested fix: instead of embedding video data inline, link to the video file using a relative path:
```python
extras.append(html_extras.html(
    f'<details><summary>Video</summary>'
    f'<video controls><source src="videos/{test_name}.webm" type="video/webm"></video></details>'
))
```
This works when the HTML report and `videos/` directory are served from the same artifact root (which your `pytest_sessionfinish` already sets up).

File: tests/suites/ui/test_sources.py [hunk 5]
```diff
@@ -0,0 +1,846 @@
...
+                session.delete(f"{gateway_url}/cost-management/v1/sources/{source['id']}")
+
+
+# =============================================================================
+# Fixtures
+# =============================================================================
+
+
+@pytest.fixture(scope="module")
+def sources_api_session(keycloak_config, gateway_url) -> requests.Session:
```
Comment (jordigilh) [inline tests/suites/ui/test_sources.py]: **[MEDIUM — QE]** Module-scoped fixture uses a single JWT token that may expire.

`obtain_jwt_token` is called once at module scope. Keycloak access tokens typically expire after 5 minutes. With 6 e2e tests that include video recording, wizard flows, and network waits, the total module runtime can easily exceed that.

If the token expires mid-run, later tests will fail on API cleanup (the `cleanup_source_by_name` calls in `finally` blocks) and the `api_created_source` fixture — producing confusing failures unrelated to the UI.

Consider using `scope="function"` or implementing token refresh logic.

File: tests/suites/ui/test_sources.py [hunk 6]
```diff
@@ -0,0 +1,846 @@
...
+def wait_for_integrations_load(page: Page, timeout_ms: int = 15000) -> None:
+    """Wait for the Integrations page to finish loading.
+    
+    The page is considered loaded when either:
+    - A table with sources is visible
+    - The empty state card is visible
+    - The loading spinner disappears
+    """
+    # Wait for loading spinner to disappear (if present)
+    spinner = page.locator(".pf-v6-c-spinner, .pf-c-spinner")
```
Comment (jordigilh) [inline tests/suites/ui/test_sources.py]: **[MEDIUM — UX/Maintainability]** PatternFly v6 CSS class selectors (`.pf-v6-c-*`) are fragile.

This file makes heavy use of PF6-specific class selectors (`.pf-v6-c-spinner`, `.pf-v6-c-table`, `.pf-v6-c-wizard`, `.pf-v6-c-modal-box`, `.pf-v6-c-backdrop`, etc.). A PatternFly major version bump would break all of these.

I see you already use `data-ouia-component-id` in some places (e.g., `sources-empty-add-openshift-card`) which is the right approach. Consider migrating the remaining selectors to OUIA IDs or `[role=...]` selectors where available. Not a blocker, but worth tracking as tech debt.

File: tests/suites/ui/conftest.py [hunk 7]
```diff
@@ -15,6 +55,24 @@
 from utils import get_route_url
 
 
+# =============================================================================
+# Reporting Configuration
+# =============================================================================
+
+# Video recording mode: "off", "on", "retain-on-failure"
+# Default to "retain-on-failure" - videos are large, only keep for failures
+VIDEO_MODE = os.environ.get("PLAYWRIGHT_VIDEO", "retain-on-failure")
```
Comment (jordigilh) [inline tests/suites/ui/conftest.py:85]: **[MEDIUM — SRE]** Default artifact capture settings will increase CI storage over time.

Current defaults: `VIDEO_MODE="retain-on-failure"`, `TRACE_MODE="on"` (all tests), `SCREENSHOT_MODE="on"` (all tests). With 6 tests, a clean run produces ~20-30MB of artifacts (traces + screenshots). This is fine now but will grow linearly with test count.

Consider:
1. Setting `TRACE_MODE` default to `"retain-on-failure"` instead of `"on"` — traces are the largest artifact type
2. Adding a note about expected artifact sizes per test count in the docstring
3. Setting up CI artifact retention policies (e.g., 30-day TTL)

File: tests/suites/ui/test_sources.py [hunk 8]
```diff
@@ -0,0 +1,846 @@
...
+
+
+# =============================================================================
+# End-to-End Workflow Tests
+# =============================================================================
+
+
+@pytest.mark.ui
+@pytest.mark.sources
+class TestIntegrationWorkflows:
```
Comment (jordigilh) [inline tests/suites/ui/test_sources.py]: **[MEDIUM — QE]** No negative/edge-case tests for wizard validation.

The current test suite covers the happy path well (6 tests, full CRUD). However, there are no tests for:
- Submitting with an empty name (is the Next button disabled?)
- Duplicate source name (what error does the wizard show?)
- Invalid or empty cluster ID
- Very long input strings

Consider adding at least one test that verifies the Next button remains disabled when required fields are empty. This catches regressions in form validation logic. Not a merge blocker — can be a follow-up.

File: tests/suites/ui/test_sources.py [hunk 9]
```diff
@@ -0,0 +1,846 @@
...
+    name: str
+    cluster_id: str
+
+
+# =============================================================================
+# Helper Functions
+# =============================================================================
+
+
+def wait_for_integrations_load(page: Page, timeout_ms: int = 15000) -> None:
```
Comment (jordigilh) [inline tests/suites/ui/test_sources.py]: **[INFO — QE, positive]** Good use of explicit wait patterns.

`wait_for_integrations_load` correctly waits for content to appear rather than using hard `time.sleep()` or `wait_for_timeout()`. The spinner-then-content pattern is robust and will reduce flakiness.

File: tests/suites/ui/conftest.py [hunk 10]
```diff
@@ -189,3 +363,228 @@ def pytest_runtest_makereport(item, call):
...
+            name_without_ext = filename[:-5]  # Remove .webm
+            if len(name_without_ext) == 32 and all(c in "0123456789abcdef" for c in name_without_ext):
+                filepath = os.path.join(videos_dir, filename)
+                try:
+                    os.remove(filepath)
+                except Exception:
+                    pass
+
+
+def pytest_sessionfinish(session, exitstatus):
```
Comment (jordigilh) [inline tests/suites/ui/conftest.py:522]: **[INFO — SRE, positive]** Good `ARTIFACT_DIR` integration.

The `pytest_sessionfinish` hook correctly copies artifacts to `ARTIFACT_DIR`, cleans up orphaned videos, and generates a browsable HTML index. This integrates well with Prow's artifact collection.

File: tests/suites/ui/conftest.py [hunk 11]
```diff
@@ -189,3 +363,228 @@ def pytest_runtest_makereport(item, call):
     outcome = yield
     rep = outcome.get_result()
     setattr(item, f"rep_{rep.when}", rep)
+
+
+@pytest.hookimpl(hookwrapper=True)
```
Comment (jordigilh) [inline tests/suites/ui/conftest.py:419]: **[INFO — FedRAMP]** Playwright traces may capture sensitive data.

Traces include DOM snapshots and network requests. If the UI displays org IDs, cluster identifiers, or cost data, these will be captured in trace files stored in CI artifact buckets.

This PR doesn't touch any deployment artifacts, so there's zero FedRAMP boundary impact on the deployed application. But for FedRAMP-scoped CI environments, ensure artifact storage buckets have appropriate access controls and retention policies.

File: tests/suites/ui/test_sources.py (hunk 2 above)
Comment (testetson22) [reply tests/suites/ui/test_sources.py]: All of that was considered and scoped deletion was the prior method, but I'll find a middle ground where we disable the creation of the CMMO created source that causes empty state issues and are less aggressive with cleanup.

File: tests/suites/ui/conftest.py (hunk 3 above)
Comment (testetson22) [reply tests/suites/ui/conftest.py]: This is a general problem; I'll document more broadly and wire up env vars.

File: tests/suites/ui/test_sources.py (hunk 5 above)
Comment (testetson22) [reply tests/suites/ui/test_sources.py]: I already handle this in API-based tests; refactoring and consolidating will fix it.

**Comment by testetson22** [comment]
I think I got all of the feedback. Added **API Test Code Consolidation** to the PR to address token/login issues that were present through the tests and had a bit of duped code; addresses token scoping issue and a bit more. Created FLPATH-4173 for negative tests.

**Review by chadcrum — APPROVED** [review]
Looks good. Thomas addressed the main concerns — token scoping consolidation, cleanup approach, and FLPATH-4173 for negative tests. Remaining threads are outdated from the latest push. /lgtm

**Review by chadcrum — APPROVED** [review]
